api/src/main/java/org/pmiops/workbench/notebooks/NotebooksServiceImpl.java

```diff
@@ -1,126 +1,151 @@
+        .getWorkspace().getBucketName();
 
-    WorkbenchConfig config = workbenchConfigProvider.get();
-    return new ClusterRequest()
-        .labels(ImmutableMap.of(
-            CLUSTER_LABEL_AOU, "true",
-            CLUSTER_LABEL_CREATED_BY, userEmail))
-        .defaultClientId(config.server.oauthClientId)
-        .jupyterUserScriptUri(config.firecloud.jupyterUserScriptUri)
-        .userJupyterExtensionConfig(new UserJupyterExtensionConfig()
-            .nbExtensions(ImmutableMap.of(
-                "playground-extension", config.firecloud.jupyterPlaygroundExtensionUri))
-            .serverExtensions(ImmutableMap.of("jupyterlab", "jupyterlab"))
-            .combinedExtensions(ImmutableMap.<String, String>of())
-            .labExtensions(ImmutableMap.<String, String>of()))
-        .machineConfig(new MachineConfig()
-            .masterDiskSize(Optional.ofNullable(clusterOverride.masterDiskSize).orElse(20 /* GB */))
-            .masterMachineType(Optional.ofNullable(clusterOverride.machineType).orElse("n1-standard-2")));
+    return cloudStorageService.getBlobList(bucketName, NOTEBOOKS_WORKSPACE_DIRECTORY).stream()
+        .filter(blob -> NOTEBOOK_PATTERN.matcher(blob.getName()).matches())
+        .map(blob -> blobToFileDetail(blob, bucketName))
+        .collect(Collectors.toList());
   }
 
-  @Override
-  public Cluster createCluster(String googleProject, String clusterName) {
-    ClusterApi clusterApi = clusterApiProvider.get();
-    User user = userProvider.get();
-    return retryHandler.run((context) ->
-        clusterApi.createClusterV2(googleProject, clusterName,
-            createFirecloudClusterRequest(user.getEmail(), user.getClusterConfigDefault())));
+  private FileDetail blobToFileDetail(Blob blob, String bucketName) {
+    String[] parts = blob.getName().split("/");
+    FileDetail fileDetail = new FileDetail();
+    fileDetail.setName(parts[parts.length - 1]);
+    fileDetail.setPath("gs://" + bucketName + "/" + blob.getName());
+    fileDetail.setLastModifiedTime(blob.getUpdateTime());
+    return fileDetail;
   }
 
   @Override
-  public void deleteCluster(String googleProject, String clusterName) {
-    ClusterApi clusterApi = clusterApiProvider.get();
-    retryHandler.run((context) -> {
-      clusterApi.deleteCluster(googleProject, clusterName);
-      return null;
-    });
+  public FileDetail copyNotebook(String fromWorkspaceNamespace, String fromWorkspaceName, String fromNotebookName,
+      String toWorkspaceNamespace, String toWorkspaceName, String newNotebookName) {
+    GoogleCloudLocators fromNotebookLocators = getNotebookLocators(fromWorkspaceNamespace, fromWorkspaceName, fromNotebookName);
+    GoogleCloudLocators newNotebookLocators = getNotebookLocators(toWorkspaceNamespace, toWorkspaceName, newNotebookName);
+
+    if (!cloudStorageService.blobsExist(Collections.singletonList(newNotebookLocators.blobId)).isEmpty()) {
```

Comment:
This should happen after the access check to avoid unnecessary exposure of information